File: packages/plugin-obsessiondb/src/auth/credentials.ts
import { mkdir, readFile, writeFile, unlink, chmod } from 'node:fs/promises'
import { dirname, join } from 'node:path'
import { homedir } from 'node:os'

export interface Credentials {
  access_token: string
  base_url: string
}

export function getCredentialsPath(): string {
  const xdgConfig = process.env.XDG_CONFIG_HOME
  const configDir = xdgConfig || join(homedir(), '.config')
  return join(configDir, 'chkit', 'credentials.json')
}

export async function loadCredentials(): Promise<Credentials | null> {
  try {
    const raw = await readFile(getCredentialsPath(), 'utf8')
    const parsed = JSON.parse(raw) as unknown
    if (
      typeof parsed === 'object' &&
      parsed !== null &&
      'access_token' in parsed &&
      'base_url' in parsed &&
      typeof (parsed as Credentials).access_token === 'string' &&
      typeof (parsed as Credentials).base_url === 'string'
    ) {
      return parsed as Credentials
    }
    return null
  } catch {
    return null
  }
}

export async function saveCredentials(creds: Credentials): Promise<void> {
  const filePath = getCredentialsPath()
  const dir = dirname(filePath)
  await mkdir(dir, { recursive: true, mode: 0o700 })
  await writeFile(filePath, JSON.stringify(creds, null, 2) + '\n', { mode: 0o600 })
  // Ensure permissions even if file existed
  await chmod(filePath, 0o600)
}

export async function clearCredentials(): Promise<void> {
  try {
    await unlink(getCredentialsPath())
  } catch {
    // Already gone — no-op
  }
}

const DEFAULT_BASE_URL = 'https://console-api.obsessiondb.com'

/**
 * Resolve the ObsessionDB API base URL.
 * Priority: OBSESSIONDB_API_URL env var > stored credentials > default.
 */
export function resolveBaseUrl(stored?: string): string {
  const envValue = process.env.OBSESSIONDB_API_URL
  if (envValue && envValue.length > 0) return envValue
  if (stored && stored.length > 0) return stored
  return DEFAULT_BASE_URL
}

File: chkit_python/src/chkit_plugin_obsessiondb/credentials.py
"""ObsessionDB credentials file (XDG-compliant).

1:1 port of ``packages/plugin-obsessiondb/src/auth/credentials.ts``.

Stores ``{access_token, base_url}`` under
``$XDG_CONFIG_HOME/chkit/credentials.json`` (default ``~/.config/chkit/``).
On POSIX the file mode is ``0o600`` and the parent directory ``0o700`` —
this matches the TS implementation and protects the access token from
other local users.
"""

from __future__ import annotations

import contextlib
import json
import os
import platform
import stat
from pathlib import Path
from typing import Any

from pydantic import BaseModel, ConfigDict

DEFAULT_BASE_URL = "https://console-api.obsessiondb.com"


class Credentials(BaseModel):
    """Persisted auth state for the ObsessionDB API."""

    model_config = ConfigDict(frozen=True, extra="ignore")

    access_token: str
    base_url: str


def get_credentials_path() -> Path:
    """Return ``$XDG_CONFIG_HOME/chkit/credentials.json``.

    Falls back to ``~/.config/chkit/credentials.json`` when XDG isn't set,
    matching the TS behaviour on every supported platform.
    """
    xdg = os.environ.get("XDG_CONFIG_HOME", "").strip()
    base = Path(xdg) if xdg else Path.home() / ".config"
    return base / "chkit" / "credentials.json"


def load_credentials() -> Credentials | None:
    """Read + validate the credentials file. Returns None on any failure."""
    path = get_credentials_path()
    try:
        raw = path.read_text(encoding="utf-8")
    except OSError:
        return None
    try:
        parsed: Any = json.loads(raw)
    except json.JSONDecodeError:
        return None
    if not isinstance(parsed, dict):
        return None
    access_token = parsed.get("access_token")
    base_url = parsed.get("base_url")
    if not isinstance(access_token, str) or not isinstance(base_url, str):
        return None
    return Credentials(access_token=access_token, base_url=base_url)


def save_credentials(creds: Credentials) -> None:
    """Write the credentials file with mode 0o600 (POSIX)."""
    path = get_credentials_path()
    path.parent.mkdir(parents=True, exist_ok=True)
    if platform.system() != "Windows":
        with contextlib.suppress(OSError):
            os.chmod(path.parent, stat.S_IRWXU)
    payload = creds.model_dump()
    path.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")
    if platform.system() != "Windows":
        with contextlib.suppress(OSError):
            os.chmod(path, stat.S_IRUSR | stat.S_IWUSR)


def clear_credentials() -> bool:
    """Delete the credentials file. Returns True if it existed and was removed."""
    path = get_credentials_path()
    try:
        path.unlink()
    except FileNotFoundError:
        return False
    except OSError:
        return False
    return True


def resolve_base_url(stored: str | None = None) -> str:
    """``OBSESSIONDB_API_URL`` env > stored value > default. Matches TS priority."""
    env_value = os.environ.get("OBSESSIONDB_API_URL", "").strip()
    if env_value:
        return env_value
    if stored:
        return stored
    return DEFAULT_BASE_URL
